### ISSUE TITLE
How to install Saleor on Ubuntu 16.04

### ISSUE THREAD

**Issue by marabqa** [mirumee/saleor#2108]
Hello,
I'm trying to install and configure Saleor on Ubuntu but whenever do the installation getting an error.
Could you please help me with the right steps how to do the installation. I tried many ways to install it but I could not able.

Looking the right way and straightforward on installing the Saleor.

**Comment by Pacu2** [comment]
Hi @marabqa,
You can find installation docs for Linux here: http://saleor.readthedocs.io/en/latest/gettingstarted/installation-linux.html
Have you tried them?

Apart from that, we're not able to help you unless you provide a complete error log.

**Comment by marabqa** [comment]
Hello, 
Yes I did the same but it wasn't worked, I got a lot of error when I tryied to did this stpes (pip install -r requirements.txt). I will share with you the error soon

**Comment by patrys** [comment]
Please make sure you're using Python 3, Saleor will not work with Python 2.7 which is the default version on old Ubuntu systems.

**Comment by marabqa** [comment]
Hello,
I know about this before I proceed this step I updated Python to v3. I will share with you the error soon.

**Comment by marabqa** [comment]
I'm stuck now at this step creating a virtual environment, I created a virtual environment 

The virtual environment was not created successfully because ensurepip is not
available.  On Debian/Ubuntu systems, you need to install the python3-venv
package using the following command.

    apt-get install python3-venv --------------------------- **This is already installed.........**

You may need to use sudo with that command.  After installing the python3-venv
package, recreate your virtual environment.

Failing command: ['/root/my_env/bin/python3', '-Im', 'ensurepip', '--upgrade', '--default-pip']

**Comment by marabqa** [comment]
Update......
I did what mentioned above and installed ensurepip but still getting the same message that you have to install apt-get install python3-venv  when I tried to do the same again getting the same message

**Comment by marabqa** [comment]
How could I fix that, so I can use only 3.4 instead of 2.x

**Comment by patrys** [comment]
It seems to be a problem specific to how Ubuntu decided to package Python 3 (which is quite unusual). Please see the following question and answer on Stack Overflow: [How to use virtualenv with python3.6 on ubuntu 16.04?
](https://stackoverflow.com/questions/47822740/how-to-use-virtualenv-with-python3-6-on-ubuntu-16-04)

I'm going to close this as it seems to be a problem of the platform rather than one related to Saleor.

**Comment by marabqa** [comment]
The cause has not solved yet and you should not close it without informing me.

**Comment by marabqa** [comment]
The link that you provided does not help me at all I checked before and still facing the same issue, what it will be another alternative option other than using Ubuntu...

**Comment by patrys** [comment]
@marabqa I'm not sure how the Saleor team can help you as what you're struggling with is getting Python 3 to work as expected. I think the Ubuntu forums may be a better place to get your system to the point where at least Python 3 works and you can create a working virtualenv. We can take over from there and provide help specific to Saleor.

For example there is a problem related to [how the `prices` library behaves on Python 3.5](https://github.com/mirumee/prices/issues/30) which I believe is the version officially provided by Ubuntu 16.04.

**Comment by marabqa** [comment]
Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-ATQrIK/cairosvg/setup.py", line 24, in <module>
        with open(init_path, 'r', encoding='utf-8') as fd:
    TypeError: 'encoding' is an invalid keyword argument for this function

    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-ATQrIK/cairosvg/

**Comment by Pacu2** [comment]
@marabqa Unfortunately we are not able to cover set up guide for every Linux system on earth(setup might vary for each distribution, version of it, release etc.), we are focusing on providing support for our product only, but we are glad to provide some tips if we can see what could possibly go wrong.

You are trying to run Saleor with wrong version of Python, as we said, Saleor supports Python 3.4 or newer, but you're using Python 2.x, please try to resolve that problem and let us know if they're any installation problems with Saleor.

**Comment by marabqa** [comment]
When you try to install Node.js it will be installing python 2.7 as this is required with Node.js. Please consider that the latest version of Ubuntu it will be Python 3.5 so that you will have two version on the same box. 

The question here how to install Node.js without Python 2.7

**Comment by NyanKiyoshi** [comment]
Note that you don't need to install the virtualenv for python3. At least in my case I never care about virtualenv being Python2 or 3. It an alias for `virtualenv -p /usr/bin/python3`.

I think that from Ubuntu 14, Python3 is the default Python version for almost every python thingy though.

---

Here a quick guide,

You may need to install a few packages if your environment is brand new, luckily I have the list right now for Ubuntu (may need to add a few things, I don't promise there is everything):

```bash
apt install python-dev libffi-dev python-psycopg2 libpq-dev libjpeg-dev zlib1g-dev python libcairo2-dev gir1.2-pango virtualenv
```


So, you can run `virtualenv -p /usr/bin/python3 env`, which would create a virtual env for python3 (it's the 3.5 on Ubuntu 16 as of now).

Then, you need to use Python, which is your Python 3 environment from your environment (shell session).
For that, run `source env/bin/activate`

Then, you need to install the dependencies through Pip. So run `pip3 -r replace_with_saleor_path/requirements.txt`

Now, you need to migrate the db (make sure you configured PostgreSQL and the server is running), `replace_with_saleor_path/manage.py migrate`.

You then need to set the `SECRET_KEY` environment variable for saleor, for that, run: `export SECRET_KEY='not so secret anymore'`.

You can populate the DB with example data now. To do that, run: `replace_with_saleor_path/manage.py populatedb --createsuperuser`.

And now, you can start the Saleor development server with `replace_with_saleor_path/manage.py runserver`.

---

To summarize:

- You need a virtualenv, with Python3.4+;
- You need to run Saleor with this Python virtualenv in your environment;
- You need to set the secret key in your environment for Saleor to work correctly;
- You need to make sure the PostgreSQL server is running and correctly configured.

**Comment by marabqa** [comment]
Here is my latest update.... what you have mentioned above is worked for me without any errors.
Now I'm able to use Saleor and I would like to thank you for providing useful details.

**Comment by NyanKiyoshi** [comment]
Awesome!

**Comment by tsoisauce** [comment]
Try 

Python 3 
> pip3 install -r requirements.txt

Python 3.6
> pip3.6 install -r requirements.txt

if you use pip install -r requirements.txt  it will default to Python 2.7 which Saleor needs Python 3.4 or later. Default Python 3 is 3.5.7.
